[progress] httpClient changes to support visualizing progress in bndtools

gamerson opened the pull request.

bjhargrave commented on biz.aQute.repository/src/aQute/bnd/jpm/Repository.java:
```diff
@@ -1606,6 +1606,7 @@ protected void fireBundleAdded(File file) throws IOException {
 	public void setRegistry(Registry registry) {
 		this.registry = registry;
 		this.httpClient = registry.getPlugin(HttpClient.class);
+		this.httpClient.setRegistry(registry);
```
This seems like a threading issue. The getPlugin returns the one? HttpClient object and then the caller set a registry in it. What happens when there are multiple callers to get the HttpClient plugin? Either on this thread or across multiple threads?

gamerson replied:
Hmm, right, that could be trouble, I checked and turns out this isn't even needed, sending a new commit.

bjhargrave commented on biz.aQute.repository/src/aQute/bnd/repository/maven/pom/provider/BndPomRepository.java:
```diff
@@ -58,6 +58,7 @@ public synchronized void init() {
 			inited = true;
 			Workspace workspace = registry.getPlugin(Workspace.class);
 			HttpClient client = registry.getPlugin(HttpClient.class);
+			client.setRegistry(workspace);
```
Remove this since it is set in the workspace.

bjhargrave commented on biz.aQute.repository/src/aQute/bnd/repository/maven/provider/MavenBndRepository.java:
```diff
@@ -422,6 +422,7 @@ private synchronized void init() {
 				return;
 
 			client = registry.getPlugin(HttpClient.class);
+			client.setRegistry(registry);
```
Remove this since it is set in the workspace.

bjhargrave commented on biz.aQute.repository/src/aQute/bnd/repository/p2/provider/P2Repository.java:
```diff
@@ -56,6 +56,7 @@ private synchronized P2Indexer getP2Index() {
 	P2Indexer getP2Index0() {
 		this.workspace = registry.getPlugin(Workspace.class);
 		HttpClient client = registry.getPlugin(HttpClient.class);
+		client.setRegistry(this.workspace);
```
Remove this since it is set in the workspace.

gamerson commented:
Can I get a LGTM from other committers?  I'd like to get this available in bnd snapshots so that its easier to get more testing from other users on the bndtools side of this issue https://github.com/bndtools/bndtools/pull/1393

gamerson commented:
I've rebased and also added a new commit that adds a class called BndProgress that can be used in bnd workspace settings.gradle file in order to add progress to the console using Gradle's ProgressLogger framework.  See screenshot

<img width="1293" alt="screen shot 2016-06-15 at 10 42 28 pm" src="https://cloud.githubusercontent.com/assets/595221/16104892/bfb8f644-334a-11e6-8e39-fc627ce5a28a.png">

In order to use this, you need to modify your settings.gradle to add the BndProgress to the workspace.

```
/* Initialize the bnd workspace */
Workspace.setDriver(Constants.BNDDRIVER_GRADLE)
Workspace.addGestalt(Constants.GESTALT_BATCH, null)
def workspace = new Workspace(rootDir, bnd_cnf)
workspace.addBasicPlugin(new BndProgress(gradle))
```

Note: only the last line is new and the first three are what is in the existing osgi/workspace template.

pkriens commented:
I want to merge this!!!

gamerson commented:
Done!
